server: use Open instead of NewPebble

RaduBerinde opened the pull request:
Clean up `server.CreateEngines` to use `Open` instead of `NewPebble` in all cases.

Release note: none
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97332)
<!-- Reviewable:end -->

jbowens reviewed (approved):
Thanks for doing this! Left a few comments that are future improvements that can definitely be deferred.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-:-NOpZ4DaCnWiWx52FyBR:byowyiw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @RaduBerinde)

___
*[`pkg/server/config.go` line 781 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpWOdA1OMGYhLGrL07:-NOpWOdA1OMGYhLGrL08:bavtenb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L781)):*
> ```Go
> 			}
> 			addCfgOpt(storage.MaxSize(sizeInBytes))
> 			addCfgOpt(storage.CacheSize(cfg.CacheSize))
> ```

I think both branches can use `storage.Caches` (lifting it into the `storageConfigOpts` declaration). We unconditionally construct the caches using `cfg.CacheSize` up above. 
___
*[`pkg/server/config.go` line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpXC834DajfQyxuGEP:-NOpXC834DajfQyxuGEQ:b61ikc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L809)):*
> ```Go
> 			addCfgOpt(storage.Caches(pebbleCache, tableCache))
> 			addCfgOpt(storage.MaxOpenFiles(int(openFileLimitPerStore)))
> 			addCfgOpt(storage.MaxWriterConcurrency(2))
> ```

are we able to lift most of these into the shared `ConfigOption` declaration? I think it's a bit of a wart that these were ever applied only for filesystem engines, since even tests using  an in-memory VFS should respect open file limits, etc. Can be a TODO.
___
*[`pkg/server/config.go` line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpX6f58hJW3thfm0rB:-NOpX6f61cemktulaBB5:b-t32mao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L811)):*
> ```Go
> 			addCfgOpt(storage.MaxWriterConcurrency(2))
> 			if sharedStorage != nil {
> 				addCfgOpt(storage.SharedStorage(sharedStorage))
> ```

looks like `sharedStorage` is added twice
___
*[`pkg/server/config.go` line 815 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpY2i23K3kujjVWMfQ:-NOpY2i23K3kujjVWMfR:bupd5jx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L815)):*
> ```Go
> 			// If the spec contains Pebble options, set those too.
> 			if spec.PebbleOptions != "" {
> 				addCfgOpt(storage.PebbleOptions(spec.PebbleOptions, &pebble.ParseHooks{
> ```

same with this too—we should be able to override Pebble options on new in-memory engines too.
___
*[`pkg/storage/open.go` line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpVGEwD6BGM-e4-34t:-NOpVGEwD6BGM-e4-34u:b-pkt2ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/storage/open.go#L176)):*
> ```Go
> // PebbleOptions configures Pebble-specific options in the same format as a
> // Pebble OPTIONS file but treating any whitespace as a newline:
> // (Eg, "[Options] delete_range_flush_delay=2s flush_split_bytes=4096")
> ```

this comment (and I think the one on `StoreSpec.PebbleOptions` too) is a little misleading. Once a store spec has been constructed through `NewStoreSpec`, the `StoreSpec.PebbleOptions` field's delimiter whitespace should have already been transformed into newlines. Down below `cfg.Opts.Parse` will require the input fields be newline-delimited.
___
*[`pkg/storage/pebble.go` line 870 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpYUp56To4i48iXtrp:-NOpYUp56To4i48iXtrq:btf48cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/storage/pebble.go#L870)):*
> ```Go
> 
> // NewPebble creates a new Pebble instance, at the specified path.
> // Do not use directly (except in test); use Open instead.
> ```

should we replace the remaining test call sites with `Open` calls too and unexport `NewPebble`? I think we would also be able to get rid of the exported `PebbleConfig` type too (either unexporting it or deleting it and moving its fields directly onto `engineConfig`)


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
server: use Open instead of NewPebble

Clean up `server.CreateEngines` to use `Open` instead of `NewPebble`.

Release note: none
Epic: none

RaduBerinde reviewed (commented):
TFTR! I want to keep the current behavior unchanged here. I am leaving TODOs though.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-:-NOpbCCK1UC-qZGfqGN2:bwgp5bv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @jbowens)

___
*[`pkg/server/config.go` line 781 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpWOdA1OMGYhLGrL07:-NOpb7S2BbeRzUomdPdC:b-fgfq42) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L781)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think both branches can use `storage.Caches` (lifting it into the `storageConfigOpts` declaration). We unconditionally construct the caches using `cfg.CacheSize` up above. 
</blockquote></details>

We only construct `tableCache` if we have non-in-memory stores..
___
*[`pkg/server/config.go` line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpXC834DajfQyxuGEP:-NOpdkeN4jyn6EgKBjIp:bkd5ega) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L809)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

are we able to lift most of these into the shared `ConfigOption` declaration? I think it's a bit of a wart that these were ever applied only for filesystem engines, since even tests using  an in-memory VFS should respect open file limits, etc. Can be a TODO.
</blockquote></details>

Added a TODO
___
*[`pkg/server/config.go` line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpX6f58hJW3thfm0rB:-NOpcJe07ZxPJRKnzINt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L811)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

looks like `sharedStorage` is added twice
</blockquote></details>

Done.
___
*[`pkg/server/config.go` line 815 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpY2i23K3kujjVWMfQ:-NOpcSqG2FoaiY2zmnUN:b-okdfba) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L815)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

same with this too—we should be able to override Pebble options on new in-memory engines too.
</blockquote></details>

Left TODO
___
*[`pkg/storage/open.go` line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpVGEwD6BGM-e4-34t:-NOpdOBJ1WbULsmJ6In_:b-c9sid4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/storage/open.go#L176)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

this comment (and I think the one on `StoreSpec.PebbleOptions` too) is a little misleading. Once a store spec has been constructed through `NewStoreSpec`, the `StoreSpec.PebbleOptions` field's delimiter whitespace should have already been transformed into newlines. Down below `cfg.Opts.Parse` will require the input fields be newline-delimited.
</blockquote></details>

Thanks for pointing that out, fixed.
___
*[`pkg/storage/pebble.go` line 870 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpYUp56To4i48iXtrp:-NOpdY0J5zBKtU2CdCpN:bjj4t6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/storage/pebble.go#L870)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

should we replace the remaining test call sites with `Open` calls too and unexport `NewPebble`? I think we would also be able to get rid of the exported `PebbleConfig` type too (either unexporting it or deleting it and moving its fields directly onto `engineConfig`)
</blockquote></details>

It would be nice but there is one in particular which I found problematic, in the metamorphic test which randomizes the pebble config directly. Anyway, it's less valuable to clean up the test code, as long as the production path is clean.


<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-:-NOqC_TA5-vOTDn7fl65:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @RaduBerinde)

___
*[`pkg/server/config.go` line 781 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpWOdA1OMGYhLGrL07:-NOqCGwE0mnTPV0tuBV4:b-nq4nos) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/server/config.go#L781)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We only construct `tableCache` if we have non-in-memory stores..
</blockquote></details>

Yeah, that seems like a bug. We'll end up creating separate table caches for each in-memory store. 


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/storage/pebble.go` line 870 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpYUp56To4i48iXtrp:-NOqGZH_E00ZrFtMxujU:bp5uavu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/storage/pebble.go#L870)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It would be nice but there is one in particular which I found problematic, in the metamorphic test which randomizes the pebble config directly. Anyway, it's less valuable to clean up the test code, as long as the production path is clean.
</blockquote></details>

I looked closer and in the metamorphic test, the options are not randomized (not sure why I thought that), they come from a parsable string. I'll try to address the tests in a separate PR.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/storage/pebble.go` line 870 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97332#-NOpYUp56To4i48iXtrp:-NOqGq_i99JoSjjxma54:bigy04d) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a2938c72fa200df6565c42d9ed9c9c14a78ca9a/pkg/storage/pebble.go#L870)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I looked closer and in the metamorphic test, the options are not randomized (not sure why I thought that), they come from a parsable string. I'll try to address the tests in a separate PR.
</blockquote></details>

Oh, no, there is also randomization. But we should be able to convert to storage options, will just need some work.

https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/metamorphic/options.go#L106


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
